includes/payments/actions.php

```diff
@@ -348,3 +348,108 @@ function edd_cleanup_stats_transients() {
 
 }
 add_action( 'edd_daily_scheduled_events', 'edd_cleanup_stats_transients' );
+
+/**
+ * Process an attempt to complete a recoverable payment.
+ *
+ * @since  2.7
+ * @return void
+ */
+function edd_recover_payment() {
+
+	$payment = new EDD_Payment( $_GET['payment_id'] );
+	if ( $payment->ID !== (int) $_GET['payment_id'] ) {
+		return;
+	}
+
+	if ( ! $payment->is_recoverable() ) {
+		return;
+	}
+
+	// Empty out the cart.
+	edd_empty_cart();
+
+	// Recover any downloads.
+	foreach ( $payment->cart_details as $download ) {
+		edd_add_to_cart( $download['id'], $download['item_number']['options'] );
+
+		// Recover any item specific fees.
+		if ( ! empty( $download['fees'] ) ) {
+			foreach ( $download['fees'] as $fee ) {
+				EDD()->fees->add_fee( $fee );
+			}
+		}
+	}
+
+	// Recover any global fees.
+	foreach ( $payment->fees as $fee ) {
+		EDD()->fees->add_fee( $fee );
+	}
+
+	// Recover any discounts.
+	if ( 'none' !== $payment->discounts && ! empty( $payment->discounts ) ){
+		$discounts = ! is_array( $payment->discounts ) ? explode( ',', $payment->discounts ) : $payment->discounts;
+
+		foreach ( $discounts as $discount ) {
+			edd_set_cart_discount( $discount );
+		}
+	}
+
+	EDD()->session->set( 'edd_resume_payment', $payment->ID );
+
+	$redirect_args = array( 'payment-mode' => $payment->gateway );
+	$redirect      = add_query_arg( $redirect_args, edd_get_checkout_uri() );
+	wp_redirect( $redirect );
+	exit;
+}
+add_action( 'edd_recover_payment', 'edd_recover_payment' );
+
+/**
+ * If the payment trying to be recovered has a User ID associated with it, we need them to login.
+ *
+ * @since  2.7
+ * @return void
+ */
+function edd_recovery_force_login_fields() {
+	$resuming_payment = EDD()->session->get( 'edd_resume_payment' );
+	if ( $resuming_payment ) {
+		$payment = new EDD_Payment( $resuming_payment );
+		if ( ! empty( $payment->user_id ) && ( ! is_user_logged_in() || $payment->user_id != get_current_user_id() ) ) {
```

Comment:
I think maybe you misunderstood my comment.

With the current iteration it is possible to see a log in form while _already_ logged in. That should not happen.